plugins/pulp_rpm/plugins/importers/yum/pulp_solv.py

```diff
@@ -0,0 +1,406 @@
+    :return: the solvable created.
+    :rtype: solv.Solvable
+    """
+    solvable = solv_repo.add_solvable()
+    for attribute_factory in attribute_factories:
+        attribute_factory(solvable, unit)
+    return solvable
+
+
+def unit_solvable_converter_factory(*attribute_factories):
+    return lambda solv_repo, unit: unit_solvable_converter(solv_repo, unit, *attribute_factories)
+
+
+rpm_unit_solvable_factory = unit_solvable_converter_factory(
+    # An RPM content unit nests dependencies in a dict format
+    # Would be provided by pulp_rpm
+    plain_attribute_factory('name'),
+    evr_attribute,
+    plain_attribute_factory('arch'),
+    plain_attribute_factory('vendor'),
+    rpm_dependency_attribute_factory('requires'),
+    rpm_dependency_attribute_factory('conflicts'),
+    rpm_dependency_attribute_factory('provides'),
+    rpm_dependency_attribute_factory('obsoletes'),
+    rpm_dependency_attribute_factory('recommends'),
+    rpm_dependency_attribute_factory('suggests'),
+    rpm_dependency_attribute_factory('supplements'),
+    rpm_dependency_attribute_factory('enhances'),
+    rpm_filelist_conversion,
+)
+
+
+class UnitSolvableMapping(object):
+    def __init__(self, pool, type_factory_mapping):
+        self.type_factory_mapping = type_factory_mapping
+        self.pool = pool
+        self.mapping = {}
+        self.repos = {}
+
+    def _register(self, unit, solvable):
+        self.mapping.setdefault(solvable.id, unit['id'])
+        self.mapping.setdefault(unit['id'], solvable)
+        _LOGGER.debug('Loaded unit %(u)s as %(s)s', {'u': unit['id'], 's': solvable})
+
+    def add_repo_units(self, units, repo_name, installed=False):
+        repo = self.repos.get(repo_name)
+        if not repo:
+            repo = self.repos.setdefault(repo_name, self.pool.add_repo(repo_name))
+            repodata = repo.add_repodata()
+        else:
+            repodata = repo.first_repodata()
+
+        for unit in units:
+            try:
+                factory = self.type_factory_mapping[unit['_content_type_id']]
+            except KeyError as err:
+                raise ValueError('Unsupported unit type: {}', err)
+            solvable = factory(repo, unit)
+            self._register(unit, solvable)
+
+        if installed:
+            self.pool.installed = repo
+
+        repodata.internalize()
+
+    def get_unit_id(self, solvable):
+        return self.mapping.get(solvable.id)
+
+    def get_solvable(self, unit):
+        return self.mapping.get(unit['id'])
+
+
+class Solver(object):
+    type_factory_mapping = {
+        'rpm': rpm_unit_solvable_factory,
+    }
+    rpm_key_fields = list(models.RPM.unit_key_fields)
+    rpm_fields = set([
+        'name',
+        'version',
+        'release',
+        'epoch',
+        'arch',
+        'vendor',
+        'provides',
+        'requires',
+        'files',
+        '_content_type_id',
+        'id',
+    ])
+    rpm_exclude_fields = set([
```

Comment:
the thing is, `RPM.objects.only('provides').as_pymongo()` doesn't expand the list of provides objects. I couldn't get it working in any other way than `RPM.objects.exclude(...).as_pymongo()` :-/